### PR TITLE
[HIP] Fix GlobalMemory launching mechanism

### PR DESCRIPTION
**Pull request by Rombur**
Code is easier to review if you hide white spaces.

Related to #5719

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP_Instance.hpp [hunk 1]
```diff
@@ -133,6 +135,7 @@ class HIPInternal {
   // Resizing of reduction related scratch spaces
   size_type *scratch_space(const std::size_t size);
   size_type *scratch_flags(const std::size_t size);
+  size_type *scratch_functor(const std::size_t size) const;
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Instance.hpp:138]: Why pretend it is const if all other aren't marked as const and it forces you to make member data mutable?

File: core/src/HIP/Kokkos_HIP_Instance.cpp [hunk 2]
```diff
@@ -296,15 +318,18 @@ void HIPInternal::finalize() {
     RecordHIP::decrement(RecordHIP::get_record(m_scratchFlags));
     RecordHIP::decrement(RecordHIP::get_record(m_scratchSpace));
 
-    for (int i = 0; i < m_n_team_scratch; ++i) {
-      if (m_team_scratch_current_size[i] > 0)
-        Kokkos::kokkos_free<Kokkos::HIPSpace>(m_team_scratch_ptr[i]);
-    }
+    if (m_scratchFunctorSize > 0)
+      RecordHIP::decrement(RecordHIP::get_record(m_scratchFunctor));
+  }
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Instance.cpp:323]: Did you mean to nest in the `if (nullptr != m_scratchSpace || nullptr != m_scratchFlags)` block?

File: core/src/HIP/Kokkos_HIP_Instance.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/HIP/Kokkos_HIP_Instance.hpp:138]: Now I see that is also how CUDA does it...

Comment (Rombur) [reply core/src/HIP/Kokkos_HIP_Instance.hpp:138]: The function needs to be `const` because it is called from a const object.

**Review by crtrott — APPROVED** [review]
The lock for not overwriting this happens elsewhere right?

**Comment by PhilMiller** [comment]
HPX and CUDA failures are irrelevant.

**Comment by dalg24** [comment]
> The lock for not overwriting this happens elsewhere right?

I think you don't need a lock for this one because the driver is being copied to an instance specific allocation.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**